### ISSUE TITLE
Tags no longer updated when restarting track, switching folders or hitting refresh

### ISSUE THREAD

**Issue by spxxxk** [quodlibet/quodlibet#2638]
In Filebrowser, QL used to update tags if they've been changed by an external app, after doing:
1/ right-click on folder, click on Refresh, or
2/ restarting a track (double-clicking a track, or play next then previous track), or
3/ going to another folder and back (not sure about that one)

Reproduce with:
- Play track1 with comment tag: old
- `$ operon set comment new track1`
- do 1/, 2/ or 3/
-> QL still shows old comment, Edit Tags window too. ExFalso shows the new one.

**Comment by spxxxk** [comment]
Tags are not updated after restarting QL either.
Tags are updated only if track's filename changed.

**Comment by pfps** [comment]
I think I'm seeing this in the paned browser.  I changed the album name on a bunch of tracks using exfalso.  The album name changed in the track list but not in the pane I had showing album names.  Even rebuilding the library didn't change the name in the album pane.  What change could have been done to introduce this bug?

**Comment by pfps** [comment]
What needs to be done to get the updated information into QL?

**Comment by lazka** [comment]
If this affects the current development build it's likely caused by the recent switch from Python 2 to Python 3.

**Comment by pfps** [comment]
Yes, I'm running under python3.
I think that this makes it harder to figure out what it going on, as there is no code change to look at.

**Comment by lazka** [comment]
Yeah, it could either be operon/exfalso not actually saving the changes, or QL not loading all.

A common error due to the 2->3 change has been that some things now return iterators instead of lists and iterating twice over them then does nothing the second time.

**Comment by pfps** [comment]
Maybe two bugs:

Editing from within Quodlibet doesn't affect the panes in the paned browser until restart.
Editing from outside of Quodlibet doesn't affect the songlist until restart.
Editing from outside of Quodlibet doesn't affect the panes in the paned browser until two restarts.

However, this does look like there is an attempt to run through an iterator twice. so changes get into the library the first time around, but then the other stuff isn't updated until QL is started again, which rebuilds the other stuff directly from the library.

**Comment by lazka** [comment]
@spxxxk there are currently two cases in play. if the songs are already in your main library only a library refresh will update them. If they are not in the main library the file browser will handle them and reoload etc.

Might be that you remembered the behavior from songs not in your library and now browse files already in your library?

@pfps thanks for testing what's broken. I think this is fixed in master now.

**Comment by spxxxk** [comment]
It's possible, although I also remembered that restarting a track would update that track's tag only (2/).
And that currently doesn't work even for tracks not in library.

Anyway, I'll close as I am not 100% sure.

**Comment by pfps** [comment]
Everything looks fine to me.  Internal edits show up everywhere.  External edits require a scan library action to show up.  Thanks for the quick fix.

**Comment by spxxxk** [comment]
Another example of how confusing that issue is, and why I think it didn't behave this way before:
I just used the "Fix TLEN tag" script on mp3 with 0:00 length, nothing happened and was about to fill a bug report.
Then I remembered I had to update the library. Not sure what someone who isn't aware of this issue would do.

Also, I don't know about average update time, but big libraries over network take time to update, so the fix isn't great.
